pkg/planner/core/plan_cache.go

```diff
@@ -553,7 +558,13 @@ func rebuildRange(p Plan) error {
 						return errors.New("rebuild to get an unsafe range")
 					}
 					for i := range ranges {
-						x.Handles[i] = kv.IntHandle(ranges[i].LowVal[0].GetInt64())
+						if i >= len(x.Handles) {
+							// Previous execution might have removed duplicates,
+							// so Handles are not the same length as HandleParams
+							x.Handles = append(x.Handles, kv.IntHandle(ranges[i].LowVal[0].GetInt64()))
```

Comment:
I recommend returning an unsafe-range error directly for safety. We've encountered some wrong result bugs caused by re-using wrong plans with unsafe-ranges, so better to return an error directly here.

Reply:
I just removed this, since the previous statement checks `if len(ranges) != len(x.Handles)` so it cannot happen here.

Reply:
I added a check and return an error in case `len(x.Handles) != len(x.HandleParams)`